presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java

```diff
@@ -217,88 +209,63 @@ public Page getOutput()
-                        completedBytes = endCompletedBytes;
-                        readTimeNanos = endReadTimeNanos;
-                    }
-
-                    currentPosition = 0;
+                else {
+                    // update operator stats
+                    long endCompletedBytes = pageSource.getCompletedBytes();
+                    long endReadTimeNanos = pageSource.getReadTimeNanos();
+                    operatorContext.recordGeneratedInput(endCompletedBytes - completedBytes, page.getPositionCount(), endReadTimeNanos - readTimeNanos);
+                    completedBytes = endCompletedBytes;
+                    readTimeNanos = endReadTimeNanos;
+
+                    currentOutput = pageProcessor.process(operatorContext.getSession().toConnectorSession(), page);
                 }
+                pageBuilderMemoryContext.setBytes(currentOutput.getRetainedSizeInBytes());
+            }
 
-                if (currentPage != null) {
-                    switch (processingOptimization) {
-                        case COLUMNAR: {
-                            Page page = pageProcessor.processColumnar(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
-                            currentPage = null;
-                            currentPosition = 0;
-                            return page;
-                        }
-                        case COLUMNAR_DICTIONARY: {
-                            Page page = pageProcessor.processColumnarDictionary(operatorContext.getSession().toConnectorSession(), currentPage, getTypes());
-                            currentPage = null;
-                            currentPosition = 0;
-                            return page;
-                        }
-                        case DISABLED: {
-                            currentPosition = pageProcessor.process(operatorContext.getSession().toConnectorSession(), currentPage, currentPosition, currentPage.getPositionCount(), pageBuilder);
-                            if (currentPosition == currentPage.getPositionCount()) {
-                                currentPage = null;
-                                currentPosition = 0;
-                            }
-                            break;
-                        }
-                        default:
-                            throw new IllegalStateException(String.format("Found unexpected value %s for processingOptimization", processingOptimization));
-                    }
-                }
+            return currentOutput.hasNext() ? currentOutput.next() : null;
+        }
 
-                pageSourceMemoryContext.setBytes(pageSource.getSystemMemoryUsage());
+        if (!finishing) {
+            int rowsProcessed = cursorProcessor.process(operatorContext.getSession().toConnectorSession(), cursor, ROWS_PER_PAGE, pageBuilder);
+
+            if (rowsProcessed == 0) {
+                finishing = true;
```

Comment:
The link doesn't work anymore, but I just copied this code from the previous version.  The current version has this check last so I have moved it back down.  The new diff will hopefully show this section unchanged.